File: reverseLinkedList-LC 206.py
# Time Complexity = O(n)
# Space Complexity = O(1)


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # If no or only 11 node present
        if head == None or head.next == None:
            return head
        
        
        # Initialising 3 pointers
        prev = None
        curr = head
        fast = head.next
        
        # Boundary condition
        while(fast != None):
            curr.next = prev
            prev = curr
            curr = fast
            fast = fast.next
        
        # As soon as fast == None, reverse the remaining last node
        curr.next = prev
        
        # Contains head of the reversed LL
        return curr
  
  
  
    '''  
    # Recursive code
    # TC = O(n)
    # SC = O(n)
    
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # If no or only 11 node present
        if head == None or head.next == None:
            return head
        
        
        reverse = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return reverse
        
        
    '''

File: reverseLinkedList_LC206.java


//Approach 1: use 3 pointers one to to hold prev node, curr node and next node
//TC:O(n) SC:O(1)

// class Solution {
//     public ListNode reverseList(ListNode head) {
//         if(head==null){
//             return head;
//         }   
        
//         ListNode current=head;
//         ListNode prev= null;
//         ListNode amele=null;
        
//         while(current!=null){
//             //see discussion
//             //first set the next element
//       amele= current.next;
//             //set the current ref to previous
//      current.next=prev;
//             //increement the previous
//             prev=current;
//             //increement the current to next
//             current=amele;
          
            
//         }
//         return prev;
//     }
// }

//approach2: recursion - 
//TC:O(n) SC:O(n)
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head==null || head.next==null){
            return head;
        }
   ListNode newHead = reverseList(head.next);
        //head.next.next is 5 and head is 4 at the first recursion output
            head.next.next =head;
            head.next=null;
        return newHead;
        
    }
}
